README.md
# Apple Receipt

[![Gem Version](https://badge.fury.io/rb/apple_receipt.svg)](https://badge.fury.io/rb/apple_receipt)
[![Build Status](https://travis-ci.org/koenrh/apple_receipt.svg?branch=master)](https://travis-ci.org/koenrh/apple_receipt)

This gem allows you to read and verify Apple receipts. It was originally built
to locally (server-side) verify the validity of receipts that are embedded in
[Status Update Notifications](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/Subscriptions.html#//apple_ref/doc/uid/TP40008267-CH7-SW13).
These receipts have a different format than the [documented](https://developer.apple.com/library/content/releasenotes/General/ValidateAppStoreReceipt/Chapters/ValidateLocally.html#//apple_ref/doc/uid/TP40010573-CH1-SW2)
App Store receipts you might be familiar with, which are [PKCS #7](https://tools.ietf.org/html/rfc2315)
containers with a payload (receipt data) encoded using [ASN.1](https://www.itu.int/itu-t/recommendations/rec.aspx?rec=X.690).

:warning: Note that this only covers the receipt data (signed data). You should
not rely on (local) verification for data that is in the notification object, but
not in the receipt (e.g. `notification_type`).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'apple_receipt'
```

And then execute:

    bundle

Or install it yourself as:

    gem install apple_receipt

## Usage

```ruby
require 'apple_receipt'

# Check receipt's validity (certificate chain, and signature)
receipt_raw = File.read('./receipt.txt')
receipt = AppleReceipt::Receipt.new(receipt_raw)
receipt.valid?
# => true

# Read receipt's data (data in example shortened for brevity)
receipt.purchase_info
# => {
#   "quantity"=>"1",
#   "expires_date_formatted"=>"2018-01-23 17:03:44 Etc/GMT",
#   "is_in_intro_offer_period"=>"false",
#   "is_trial_period"=>"false",
#   "item_id"=>"1190360447",
#   "app_item_id"=>"947936149",
#   "transaction_id"=>"160000408504141",
#   "web_order_line_item_id"=>"160000011000001",
#   "bid"=>"com.foo.bar",
#   "product_id"=>"com.foo.bar.monthly",
#   "purchase_date"=>"2017-12-23 17:03:44 Etc/GMT",
#   "original_purchase_date"=>"2017_12_23 17:03:53 Etc/GMT"
# }
```

## Receipts

A receipt is encoded as base64, and is formatted as a [NeXTSTEP](https://en.wikipedia.org/wiki/Property_list#NeXTSTEP)
dictionary:

```
{
  "signature" = "[base64-encoded signature]";
  "purchase-info" = "[base64-encoded purchase data]";
  "pod" = "[integer]";
  "signing-status" = "0";
}
```

### Signature

The `signature` entry contains base64-encoded binary data, which has the following
layout:

- **1 byte** - Receipt version (e.g. version 3).
- **128 bytes** (version 2) or **256 bytes** (version 3) - Signature.
- **4 bytes** - Length (in number of bytes) of the certificate.
- **N bytes** - DER-encoded certificate.

The version 2 and 3 receipt certificates are signed, respectively, by:

- **Apple iTunes Store Certification Authority** (version 2)
  - Serial: 26 (`0x1a`)
  - Subject: `C=US, O=Apple Inc., OU=Apple Certification Authority, CN=Apple iTunes Store Certification Authority`
- **Apple Worldwide Developer Relations Certification Authority** (version 3)
  - Serial: 134752589830791184 (`0x1debcc4396da010`)
  - Subject: `C=US, O=Apple Inc., OU=Apple Worldwide Developer Relations, CN=Apple Worldwide Developer Relations Certification Authority`

Both certificates chain up to:

- **Apple Root CA**
  - Serial: 2 (`0x2`)
  - Subject: `C=US, O=Apple Inc., OU=Apple Certification Authority, CN=Apple Root CA`

### Purchase info

The `purchase-info` entry contains a base64-encoded NeXTSTEP dictionary that contains
the actual receipt data (purchase info).

## Validation

First, the signing certificate is parsed from the signature binary data. The
validation of the receipt works as follows.

1. Verify that the signing certificate is valid, i.e. it is not expired, and
   it chains up to either of the bundled Apple root certificates.
2. Verify that the signature over the signed data (version number and receipt
   data) is signed by the private key that correspond to the public key that is
   in the signing certificate.

## Contributing

Bug reports and pull requests are welcome on [GitHub](https://github.com/koenrh/apple_receipt).
This project is intended to be a safe, welcoming space for collaboration, and
contributors are expected to adhere to the [Contributor Covenant](https://www.contributor-covenant.org)
code of conduct.

## License

The gem is available as open source under the terms of the [ISC License](https://opensource.org/licenses/ISC).


CHANGELOG.md
# Apple Receipt changelog

## v0.2.3

- Prepared for publication to GitHub Package Registry

## v0.2.2

- Updated README to add information about validation [\#13](https://github.com/koenrh/apple_receipt/pull/13)
- Fixed snake case keys of 'purchase information' hash [\#11](https://github.com/koenrh/apple_receipt/pull/11)

## v0.2.1

- Fixed path to bundled certificates [\#10](https://github.com/koenrh/apple_receipt/pull/10)

## v0.2.0

- Added ISC license [\#5](https://github.com/koenrh/apple_receipt/pull/5)
- Added missing dependencies ([\#3](https://github.com/koenrh/apple_receipt/pull/3)
  and [\#4](https://github.com/koenrh/apple_receipt/pull/4))
- Added support for 'version 2' receipts ([\#1](https://github.com/koenrh/apple_receipt/pull/1))


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'apple_receipt'

require 'minitest/autorun'
require 'minitest/mock'
require 'minitest/pride'
require 'minitest/spec'
require 'mocha/mini_test'


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'apple_receipt'

require 'minitest/autorun'
require 'minitest/mock'
require 'minitest/pride'
require 'minitest/spec'
require 'mocha/mini_test'


test/apple_receipt_test.rb
# frozen_string_literal: true

require 'test_helper'

class AppleReceiptTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::AppleReceipt::VERSION
  end

  def test_it_does_something_useful
    assert false
  end
end


lib/apple_receipt/validator.rb
# frozen_string_literal: true

require 'openssl'

module AppleReceipt
  # Validator allows one to check the validity of a receipt.
  class Validator
    INTERMEDIATE_CERT_MAPPING = {
      3 => 'AppleWorldwideDeveloperRelationsCertificationAuthority',
      2 => 'AppleITunesStoreCertificationAuthority'
    }.freeze

    def initialize(receipt, certificates: [])
      populate_certificate_store(receipt.version, certificates)
      @receipt = receipt
    end

    def populate_certificate_store(version, provided_certificates)
      if provided_certificates.any?
        add_certificates(provided_certificates)
      else
        add_named_certificate('AppleRootCA')
        add_named_certificate(INTERMEDIATE_CERT_MAPPING[version])
      end
    end

    def add_named_certificate(name)
      cert_path = File.expand_path("../../certificates/#{name}.cer", __dir__)
      cert_file = File.read(cert_path)
      store.add_cert(OpenSSL::X509::Certificate.new(cert_file))
    end

    def add_certificates(certificates)
      certificates.each do |cert|
        store.add_cert(cert)
      end
    end

    def valid?
      store.verify(receipt.certificate) &&
        public_key.verify(OpenSSL::Digest::SHA1.new,
                          receipt.signature, signed_data)
    end

    def public_key
      receipt.certificate.public_key
    end

    def signed_data
      [receipt.version, receipt.data].pack('CA*')
    end

    def store
      @store ||= OpenSSL::X509::Store.new
    end

    private

    attr_reader :receipt
  end
end


lib/apple_receipt/receipt_parser.rb
# frozen_string_literal: true

require 'base64'
require 'openssl'
require 'json'
require 'set'

require 'apple_receipt/next_step_parser'

module AppleReceipt
  # ReceiptParser contains helper methods to parse receipt data structures.
  module ReceiptParser
    module_function

    SIGNATURE_LENGTH_MAPPING = {
      2 => 128,
      3 => 256
    }.freeze

    def parse(input)
      receipt_hash = NextStepParser.parse(input)

      unless Set['signature', 'purchase_info'].subset?(receipt_hash.keys.to_set)
        raise ArgumentError, 'Missing required fields'
      end

      signature_decoded = Base64.decode64(receipt_hash['signature'])
      data_decoded = Base64.decode64(receipt_hash['purchase_info'])

      version, signature, receipt_cert = read_signature(signature_decoded)
      [version, signature, receipt_cert, data_decoded]
    end

    def read_signature(signature_decoded)
      sig = StringIO.new(signature_decoded)
      version = sig.read(1).unpack1('C') # 8-bit unsigned (unsigned char)

      unless SIGNATURE_LENGTH_MAPPING.key?(version)
        raise ArgumentError, "Unsupported receipt version: #{version}"
      end

      signature = sig.read(SIGNATURE_LENGTH_MAPPING[version])
      cert_size = sig.read(4).unpack1('L>') # 32-bit unsigned, big-endian
      receipt_cert = OpenSSL::X509::Certificate.new(sig.read(cert_size))

      [version, signature, receipt_cert]
    end
  end
end
